File: pathsum.java
//o(n) time and o(n) memory
//passed all leetcode case
//used approach discussed in class

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    List<List<Integer>> result;
    int target;
    public List<List<Integer>> pathSum(TreeNode root, int sum) {
        result = new ArrayList<>();
        target = sum;
        if(root ==null) return result;
        dfs(root,0,new ArrayList<>());
        return result;
    }
    private void dfs(TreeNode root, int currsum, List<Integer> temp){
        //base
        if(root ==null) return;
        currsum +=root.val;
        temp.add(root.val);
        if(root.left==null&& root.right==null) {
            if(target==currsum) result.add(new ArrayList<>(temp));
        }
        dfs(root.left,currsum,temp);
        dfs(root.right,currsum,temp);
        temp.remove(temp.size()-1);
    }
}

File: path_sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    
    vector<vector<int>> result;
    vector<vector<int>> pathSum(TreeNode* root, int targetSum) {

            vector<int> ans;
            inorder(root,targetSum,0,ans);
            return result;
        
    }

    void inorder(TreeNode* root,int targetSum,int sum,vector<int> ans){
            if(root == NULL){
                return ;
            }


            sum = sum + root->val;
            ans.push_back(root->val);
            inorder(root->left,targetSum,sum,ans);
            if(root->left == NULL && root->right == NULL){
                if(targetSum == sum){
                        return result.push_back(ans);
                }
            }

        inorder(root->right,targetSum,sum,ans);


    }
};